internal/perms/acl.go

```diff
@@ -27,6 +28,16 @@ type ACLResults struct {
 	scopeMap map[string][]Grant
 }
 
+// ListPermission objects are built off of a set of Boundary grants, and
+// contain the required data to enable database query optimization,
+// specifically for Boundary resource listing.
+type ListPermission struct {
+	ScopeId     string   // The scope id for which the permission applies.
+	ResourceIds []string // Any specific resource ids that have been referred in the grant's `id` field, if applicable.
+	OnlySelf    bool     // The grant only allows actions against the user's own resources.
+	All         bool     // We got a wildcard in the grant string's `id` field.
+}
+
```

Comment:
I am thinking that maybe something like this would make sense:

```suggestion
// Permission provides information about the specific
// resources that a user has been granted access to for a given scope, resource, and action.
type Permission struct {
	ScopeId     string   // The scope id for which the permission applies.
	Resource resrouce.Type
	Action action.Type

	ResourceIds []string // Any specific resource ids that have been referred in the grant's `id` field, if applicable.
	OnlySelf    bool     // The grant only allows actions against the user's own resources.
	All         bool     // We got a wildcard in the grant string's `id` field.
}
```

Reply:
Yep, sounds good! [bb61302](https://github.com/hashicorp/boundary/pull/2335/commits/bb61302c23afbdd45259a44f39711299a2fc450e)